Implement drop-on-canvas + linkconnectoradapter consolidation

benceruleanlu opened the pull request:
Implements droponcanvas functionality and a linkconnectoradapter refactor.

- Drop on canvas (Shift and default) integrated via LinkConnector ‘dropped-on-canvas’ with proper CanvasPointerEvent.
- LinkConnector adapter: now wraps the live canvas linkConnector (no duplicate state); added dropOnCanvas() helper.
- Tests: Playwright scenarios for Shift-drop context menu/searchbox, pinned endpoint, type prefilter, and post-selection auto-connect (browser_tests/tests/vueNodes/interactions/links/linkInteraction.spec.ts).

There are some followup PRs that will fix/refactor some more noncritical things, like the terrible slotid, the number/string nodeid confusion, etc.

https://github.com/Comfy-Org/ComfyUI_frontend/pull/5780 (snapping) <-- https://github.com/Comfy-Org/ComfyUI_frontend/pull/5898 (drop on canvas + linkconnectoradapter refactor) <-- https://github.com/Comfy-Org/ComfyUI_frontend/pull/5903 (fix reroute snapping)

benceruleanlu committed:
Add dragging test

benceruleanlu committed:
Litegraph? Never heard if it

benceruleanlu committed:
Add more test cases

Commit:
Update test expectations [skip ci]

benceruleanlu committed:
Merge branch 'bl-tests' into bl-more-slots

benceruleanlu committed:
More test cases v1

benceruleanlu committed:
review comments

benceruleanlu committed:
Merge remote-tracking branch 'origin/bl-tests' into bl-more-slots

benceruleanlu committed:
Allow moving links and support reroutes

benceruleanlu committed:
Add dragging input to input drags existing link test

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into bl-more-slots

benceruleanlu committed:
nit

benceruleanlu committed:
Support dragging from output to output

benceruleanlu committed:
Add snapshot

benceruleanlu committed:
o-o shift test

benceruleanlu committed:
Update test expectation

benceruleanlu committed:
Switch to adapter approach

benceruleanlu committed:
clean up onPointerDown

benceruleanlu committed:
Add reroute anchor tests

benceruleanlu committed:
Fix double links

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into bl-more-slots

benceruleanlu committed:
temp screenshots

benceruleanlu committed:
improve typing

benceruleanlu committed:
nit

benceruleanlu committed:
cleanup unused

benceruleanlu committed:
huh?

benceruleanlu committed:
I am the one who knocks

benceruleanlu committed:
Those who type

benceruleanlu committed:
fix bad fallback and remove logging

benceruleanlu committed:
nit

benceruleanlu committed:
nit

benceruleanlu committed:
nit

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into bl-more-slots

benceruleanlu committed:
nit

benceruleanlu committed:
refactor linkInteraction.spec.ts

benceruleanlu committed:
those who know

benceruleanlu committed:
sure

benceruleanlu committed:
get nodeid and slotkey

benceruleanlu committed:
Visually snap to node

benceruleanlu committed:
Remove debug logging

benceruleanlu committed:
Try connecting to snapped first

benceruleanlu committed:
those who know cont

benceruleanlu committed:
nit

benceruleanlu committed:
type

benceruleanlu committed:
Merge remote-tracking branch 'origin/bl-more-slots' into bl-snap

benceruleanlu committed:
fix stale

benceruleanlu committed:
refactor candidatefromnodetarget

benceruleanlu committed:
Implement caching and rAF

benceruleanlu committed:
Add tests

benceruleanlu committed:
knip

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into bl-more-slots

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into bl-more-slots

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into bl-more-slots

Commit:
Update test expectations [skip ci]

benceruleanlu committed:
Merge remote-tracking branch 'origin/bl-more-slots' into bl-snap

Commit:
Update test expectations [skip ci]

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into bl-snap

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into bl-snap

benceruleanlu committed:
Remove duplicated playwright snapshots

Commit:
Update test expectations [skip ci]

benceruleanlu committed:
ci sanity check

benceruleanlu committed:
Align links to slots in subgraphs (#5876)

When vue nodes is enabled, switch to the new getSlotPosition for
subgraph nodes as well.

This aligns links to slots in subgraphs.

Do note, getSlotPosition is actually the new standard, but we fallback
to inputNode.getInputPos since it's so core, idea is to switch over
fully to getSlotPosition without the vue flag once things are more
stable.

┆Issue is synchronized with this [Notion
page](https://www.notion.so/PR-5876-Align-links-to-slots-in-subgraphs-27f6d73d365081148b92f312a0af84a0)
by [Unito](https://www.unito.io)

benceruleanlu committed:
Add more ts files to allowDefaultProject

benceruleanlu committed:
Temporarily increase playwright timeout

benceruleanlu committed:
Temporarily regenerate screenshot expectations

benceruleanlu committed:
Export interface SlotLinkDragSession

benceruleanlu committed:
Remove slotLinkPreviewRenderer

benceruleanlu committed:
Add more test cases

benceruleanlu committed:
Make linkconnector use the canonical linkconnector

benceruleanlu committed:
add canvasPointerEvent helper

benceruleanlu committed:
Add linkDropOrchestrator

benceruleanlu committed:
Add canvas drop to useSlotLinkInteraction.ts

benceruleanlu committed:
Revert temporary playwright timeout increase

benceruleanlu committed:
Merge origin/main into bl-snap: resolve conflicts

- Snap preview target: use snapped candidate position when compatible
- Vue nodes interaction: track hovered candidate via setCandidate
- Keep PR browser test snapshots (ours) for link interactions

Commit:
[automated] Update test expectations

benceruleanlu committed:
Merge remote-tracking branch 'origin/bl-snap' into bl-linkslot-refactor

Commit:
[automated] Update test expectations

benceruleanlu commented on src/renderer/core/canvas/links/linkDropOrchestrator.ts:
```diff
@@ -0,0 +1,113 @@
+    if (cached != null) {
+      candidate.compatible = cached
+    } else {
+      const nodeId = Number(layout.nodeId)
```
this is something I'm working on to fix, it's scuffed right now but works, it's because of the terrible group node code that forces it to be number or string type

benceruleanlu replied:
resolved to string

benceruleanlu commented on src/renderer/extensions/vueNodes/composables/useSlotLinkInteraction.ts:
```diff
@@ -651,7 +637,7 @@ export function useSlotLinkInteraction({
-    app.canvas?.setDirty(true)
+    app.canvas?.setDirty(true, true)
     event.preventDefault()
     event.stopPropagation()
```
i didn't even change that line...

benceruleanlu commented on src/renderer/extensions/vueNodes/composables/useSlotLinkInteraction.ts:
```diff
@@ -383,7 +335,17 @@ export function useSlotLinkInteraction({
+
+    app.canvas?.setDirty(true, true)
   }
   const raf = createRafBatch(processPointerMoveFrame)
```
true, let me recheck perf

benceruleanlu replied:
resolved with one more cache on target

christian-byrne commented on src/renderer/core/canvas/links/linkDropOrchestrator.ts:
```diff
@@ -0,0 +1,113 @@
+
+  if (!adapter || !graph) return null
+
+  const nodeId = Number(nodeIdStr)
```
We technically support string-based node IDs if you look at the backend code - esp. for subgraphs or virtual nodes on the backend IIRC.

benceruleanlu commented on src/renderer/extensions/vueNodes/composables/useSlotLinkInteraction.ts:
```diff
@@ -360,10 +309,13 @@ export function useSlotLinkInteraction({
     let candidate: SlotDropCandidate | null = state.candidate
 
     if (hoverChanged) {
```
caN WE JUST RETURN?

benceruleanlu replied:
done

benceruleanlu committed:
Fix reroute links not snapping correctly (#5903)

Summary

- Route LGraphNode slot position queries through a single source of
truth that prefers layoutStore’s DOM‑tracked slot positions and falls
back to geometry when unavailable.
- Ensures that reroute‑origin drags snap to the same visual slot centers
used by Vue nodes when hovering compatible nodes.

What changed

- LGraphNode.getInputPos/getOutputPos now resolve via getSlotPosition(),
which:
- Returns the Vue nodes layoutStore position if the slot has been
tracked.
- Otherwise, computes from node geometry (unchanged fallback behavior).
- LGraphNode.getInputSlotPos(input) resolves index→getSlotPosition() and
retains a safe fallback when an input slot object doesn’t map to an
index.

Why

- Previously, when starting a drag from a reroute and hovering a node,
the temporary link endpoint would render toward LiteGraph’s calculated
slot position, not the Vue‑tracked slot position, causing a visible
mismatch.
- By making all slot position lookups consult the layout store first,
node hover snap rendering is consistent across slot‑origin and
reroute‑origin drags.

Impact

- No behavior change for non‑Vue nodes mode or when no tracked layout
exists — the legacy calculated positions are still used.
- Centralizes slot positioning in one place, reducing special‑case logic
and making hover/drag rendering more predictable.

https://github.com/Comfy-Org/ComfyUI_frontend/pull/5780 (snapping) <--
https://github.com/Comfy-Org/ComfyUI_frontend/pull/5898 (drop on canvas
+ linkconnectoradapter refactor) <--
https://github.com/Comfy-Org/ComfyUI_frontend/pull/5903 (fix reroute
snapping)

┆Issue is synchronized with this [Notion
page](https://www.notion.so/PR-5903-Use-layoutStore-slot-positions-for-node-slot-queries-fix-reroute-origin-node-snap-2816d73d36508184b297f46b39105545)
by [Unito](https://www.unito.io)

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into bl-linkslot-refactor

benceruleanlu commented on src/renderer/core/canvas/interaction/canvasPointerEvent.ts:
```diff
@@ -0,0 +1,74 @@
+  y: number
+}
+
+const pointerHistory = new Map<number, PointerOffsets>()
```
it's cleaned up properly right now

benceruleanlu committed:
Renamings for clarity, nodeID to standardized string, and cache guarding node.closest DOM read

benceruleanlu committed:
Rename and extractions for clarity

Commit:
[automated] Update test expectations

benceruleanlu committed:
Merge remote-tracking branch 'origin/main' into bl-linkslot-refactor

christian-byrne committed:
Merge branch 'main' into bl-linkslot-refactor

actions-user committed:
[automated] Apply ESLint and Prettier fixes

christian-byrne committed:
Merge branch 'main' into bl-linkslot-refactor

Commit:
[automated] Update test expectations

christian-byrne reviewed (approved):
LGTM!
